src/librustc_typeck/check/_match.rs

```diff
@@ -605,6 +601,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
 
         // Type check the path.
         let pat_ty = self.instantiate_value_path(segments, opt_ty, def, pat.span, pat.id);
+        self.write_ty(pat.id, pat_ty);
```

Comment:
I wanted to pull all the `write_ty`s down to one spot so that types for expressions always get written in the same place. Then I could add the adjustment in that spot.